### PR TITLE
Gamelist speedup

### PR DESCRIPTION
**Pull request by shuffle2**
As stated in the kill-emu-state PR, killing emu state isn't really required for this (it would just be nice).  
This PR is ready for comments.

### CODE REVIEW

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 1]
```diff
@@ -271,9 +243,55 @@ GameListCtrl::GameListCtrl(wxWindow* parent, const wxWindowID id, const wxPoint&
...
+      Common::SetCurrentThreadName("gamelist scanner");
+
+      if (SyncCacheFile(false))
+      {
+        // Account for changes outside the cache (just wii banners atm; could scan Ini here)
+        // Note this is on the initial load path. Further improvements could be made if updates
+        // from scan->ui threads described small changelists instead of updating the entire list
+        // at once.
+        bool cache_modified = false;
+        for (auto& file : m_cached_files)
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: I don't suppose theres a standard/STL thing that expresses intent a little better (like, "check all and tell me if at least one of them returned true", eg. `std::any_of`/ `std::accumulate`/... except that all items are always checked instead of stopping for the first item that matches)?
Otherwise, some braces please, since this is two levels deep and the inside of the `for`-loop is more than one line.

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 2]
```diff
@@ -363,24 +381,40 @@ void GameListCtrl::BrowseForDirectory()
...
+void GameListCtrl::RefreshList()
 {
   int scrollPos = wxWindow::GetScrollPos(wxVERTICAL);
   // Don't let the user refresh it while a game is running
   if (Core::GetState() != Core::State::Uninitialized)
     return;
 
-  ScanForISOs();
+  m_shown_files.clear();
+  for (auto& item : m_cached_files)
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Same down here wrt to braces (or some other C++ thing)

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 3]
```diff
@@ -363,24 +381,40 @@ void GameListCtrl::BrowseForDirectory()
...
+      if (file->IsValid())
+        m_shown_files.push_back(file);
+    }
+  }
 
   Freeze();
   ClearAll();
 
-  if (m_ISOFiles.size() != 0)
+  if (m_shown_files.size() != 0)
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: `!m_shown_files.empty()`?

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 4]
```diff
@@ -404,14 +438,13 @@ void GameListCtrl::ReloadList()
...
     SetColumnWidth(COLUMN_DUMMY, 0);
-
     for (const auto& c : m_columns)
     {
       SetColumnWidth(c.id, c.visible ? FromDIP(c.default_width + platform_padding) : 0);
     }
 
     // add all items
-    for (int i = 0; i < (int)m_ISOFiles.size(); i++)
+    for (int i = 0; i < (int)m_shown_files.size(); i++)
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Maybe replace the c-style cast while we're at it?

File: Source/Core/DolphinWX/GameListCtrl.cpp [hunk 5]
```diff
@@ -759,7 +867,7 @@ void GameListCtrl::OnKeyPress(wxListEvent& event)
   static int lastKey = 0, sLoop = 0;
   int Loop = 0;
 
-  for (int i = 0; i < (int)m_ISOFiles.size(); i++)
+  for (int i = 0; i < (int)m_shown_files.size(); i++)
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Moar c-style casts. Although this one looks a little like it could be rewritten as something else (`std::find_first`?) if the order of `m_shown_files` is the same as the ordering inside the game list; and the `GameListItem` has the ability to return the same text that `bleh` would have later. Might be wrong tho...

File: Source/Core/DolphinWX/GameListCtrl.h [hunk 6]
```diff
@@ -90,13 +95,22 @@ class GameListCtrl : public wxListCtrl
...
-  struct {
+  struct
+  {
     std::vector<int> flag;
     std::vector<int> platform;
     std::vector<int> utility_banner;
   } m_image_indexes;
 
+  // Actual backing GameListItems are maintained in a background thread and cached to file
+  static constexpr u32 CACHE_REVISION = 0;
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.h]: Do we want (have to?) keep the "last updated in PR" comment alive here, to make sure _something_ conflicts if two PRs change it at the same time? Unlikely, but never say never...

File: Source/Core/DolphinWX/ISOFile.cpp [hunk 7]
```diff
@@ -266,6 +208,37 @@ bool GameListItem::ReadPNGBanner(const std::string& path)
...
+    std::memcpy(m_image.GetData(), m_pImage.data(), m_pImage.size());
+  }
+}
+
+bool GameListItem::ReloadBannerIfNeeded()
+{
+  // Wii banners can only be read if there is a savefile,
+  // so sometimes caches don't contain banners. Let's check
+  // if a banner has become available after the cache was made.
+  if ((m_Platform == DiscIO::Platform::WII_DISC || m_Platform == DiscIO::Platform::WII_WAD) &&
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/ISOFile.cpp]: PR #5651 has `DiscIO::IsWii` which might be useful here.

File: Source/Core/DolphinWX/GameListCtrl.h (hunk 6 above)
Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.h]: We should keep it here.

File: Source/Core/DolphinWX/ISOFile.cpp (hunk 7 above)
Comment (JosJuice) [inline Source/Core/DolphinWX/ISOFile.cpp]: Sure, if that PR gets merged before this one. (If that happens, there's going to be a merge conflict due to that PR adding a call to `DiscIO::IsWii` in the old copy of this code, so it's going to be noticeable.)

File: Source/Core/Common/FileSearch.cpp [hunk 8]
```diff
@@ -52,11 +61,66 @@ std::vector<std::string> DoFileSearch(const std::vector<std::string>& exts,
...
+  return DoFileSearchNoSTL(directories, exts, recursive);
+}
+
+#else
+
+std::vector<std::string> DoFileSearch(const std::vector<std::string>& directories,
+                                      const std::vector<std::string>& exts, bool recursive)
+{
+  // Callers can use this to indicate they want all files + directories in results
+  bool accept_all = exts.size() == 0;
```
Comment (Warepire) [inline Source/Core/Common/FileSearch.cpp]: `exts.empty()`?

File: Source/Core/Common/FileSearch.cpp [hunk 9]
```diff
@@ -52,11 +61,66 @@ std::vector<std::string> DoFileSearch(const std::vector<std::string>& exts,
...
-                            [&](const File::FSTEntry& entry) { return entry.isDirectory; });
+  return DoFileSearchNoSTL(directories, exts, recursive);
+}
+
+#else
+
+std::vector<std::string> DoFileSearch(const std::vector<std::string>& directories,
+                                      const std::vector<std::string>& exts, bool recursive)
+{
+  // Callers can use this to indicate they want all files + directories in results
```
Comment (Warepire) [inline Source/Core/Common/FileSearch.cpp]: Can this comment be clearer, and go somewhere else, like in the header file?

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 1 above)
Comment (Warepire) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Do you need more than one hit, or am I just stupid? `std::find_if` should do?

Comment (JosJuice) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: This loop is doing two things at once: Updating all banners that need to be updated, and checking if the cache needs to be rewritten. If you use `std::find_if`, you'll get the problem that the loop never will update more than one banner.

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 4 above)
Comment (shuffle2) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: see next comment

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 5 above)
Comment (shuffle2) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: I tried not to touch _too_ much of existing code. Really the loop this function is in should be rewritten completely, but it's not part of this PR, and I'd rather not get lost on putting more effort into WX than necessary.

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 2 above)
Comment (shuffle2) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: done

File: Source/Core/DolphinWX/GameListCtrl.cpp (hunk 1 above)
Comment (shuffle2) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: done, but let me know if you find a good replacement for the intent, I don't know of one.

Comment (BhaaLseN) [inline Source/Core/DolphinWX/GameListCtrl.cpp]: Adding braces is good enough for me. I wasn't sure if there is any, just thought I'd put it out there.

**Comment by shuffle2** [comment]
This was split into #5620 and #5659 . #5659 replaces this.
